src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+#include "igraph_paths.h"
+#include "igraph_adjlist.h"
+#include "igraph_types.h"
+#include "igraph_memory.h"
+#include "igraph_error.h"
+#include "igraph.h"
+#include "igraph_interface.h"
+
+#include <cstring>
+#include <cmath>
+#include <map>
+#include <climits>
+#include <vector>
+#include <set>
+#include <algorithm>
+#include <iterator>
+
+/*
+ *
+ * Usage: Data structure to store index value of subsets.
+ * The index would be used in DP table
+ *
+ */
+std::map<std::set<igraph_integer_t>, igraph_integer_t> subsetMap;
+
+/*
+ *	Generates Subsets of length > 1 for list of integer values.
+ *	As algorithm computes the minimum distance between set of variables ranging from
+ *	length 2 to steiner_terminals - 1, we need these subsets.
+ *   for e.g [2,3,4] --> [ [2,3], [2,4], [3,4], [2,3,4] ]
+ *	Populates the subsetMap data structure as well.
+ *
+ */
+
+std::set<std::set<igraph_integer_t>> generateSubsets(igraph_vector_int_t steinerTerminals, igraph_integer_t n, igraph_integer_t graphsize)
+{
+	igraph_integer_t count = ((igraph_integer_t)1 << n);
+	std::set<std::set<igraph_integer_t>> allSubsets;
+	igraph_integer_t subsetIndex = graphsize;
+
+	// The outer for loop will run 2^n times to get all subset .
+	// Here variable i will act as a binary counter
+
+	for (igraph_integer_t i = 0; i < count; i++)
+	{
+		// The inner for loop will run n times , As the maximum number of elements a set can have is n
+		// This loop will generate a subset
+		std::set<igraph_integer_t> newSubset;
+		for (igraph_integer_t j = 0; j < n; j++)
+		{
+			// This if condition will check if jth bit in binary representation of  i  is set or not
+			// if the value of (i & (1 << j)) is greater than 0 , include arr[j] in the current subset
+			// otherwise exclude arr[j]
+			if ((i & ((igraph_integer_t)1 << j)) > 0)
+			{
+				newSubset.insert(VECTOR(steinerTerminals)[j]);
+			}
+		}
+
+		if (newSubset.size() > 1)
+		{
+			if (allSubsets.find(newSubset) == allSubsets.end())
+			{
+				allSubsets.insert(newSubset);
+				subsetMap.insert(std::make_pair(newSubset, subsetIndex));
+				subsetIndex++;
+			}
+		}
+	}
+	return allSubsets;
+}
+
+/*
+ *
+ * Purpose: Fetching Index of a subset from subsetMap in order to store and look-up
+ * the value of subset from DP table.
+ */
+
+igraph_integer_t fetchIndexofMapofSets(std::set<igraph_integer_t> subset)
+{
+	std::map<std::set<igraph_integer_t>, igraph_integer_t>::iterator it;
+	for (it = subsetMap.begin(); it != subsetMap.end(); ++it)
+	{
+		if (it->first == subset)
+		{
+			return it->second;
+		}
+	}
+	return -1;
+}
+
+/*
+ * Purpose: Retriving the value of subset from given index.
+ *
+ */
+
+std::set<igraph_integer_t> fetchSetsBasedonIndex(igraph_integer_t index)
+{
+	std::map<std::set<igraph_integer_t>, igraph_integer_t>::iterator it;
+	for (it = subsetMap.begin(); it != subsetMap.end(); ++it)
+	{
+		if (it->second == index)
+		{
+			return it->first;
+		}
+	}
+	return std::set<igraph_integer_t>();
```

Comment:
Same comment as with the `-1` return from `fetchIndexofMapofSets()`. Is this expected to ever be executed?